agent/app/agent_unix.go

```diff
@@ -0,0 +1,76 @@
+// +build linux
+
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package app
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	log "github.com/cihub/seelog"
+	"golang.org/x/net/context"
+)
+
+// startSigchldHandler registers a channel to receiev SIGCHLD signals for
+// the agent process. On receiving SIGCHLD, it invokes the wait4 syscall
+// so that child processes are appropriately cleaned up.
+func (agent *ecsAgent) startSigchldHandler(ctx context.Context) {
+	signals := make(chan os.Signal)
+	signal.Notify(signals, syscall.SIGCHLD)
+	go processSignal(ctx, signals)
+}
+
+func processSignal(ctx context.Context, signals <-chan os.Signal) {
+	log.Info("Starting the SIGCHLD handler")
+	for {
+		select {
+		case s := <-signals:
+			log.Debugf("Received SIGCHLD: %s", s.String())
+			go wait()
+		case <-ctx.Done():
+			log.Info("Stopping the SIGCHLD handler")
```

Comment:
Since we expect this context to be cancelled only when the agent exits, I think we should be fine with just invoking `Stop()` as that guarantees that the channel will not receive any more events as per https://golang.org/pkg/os/signal/#Stop